# Trabalho Prático do Módulo 4

## Pipeline de Dados - Enem 2019

## Atividades:

Você deverá desempenhar as seguintes atividades:

1. Extrair, de maneira programática, os dados do ENEM 2019.
2. Filtrar apenas os alunos do estado de Minas Gerais.
3. Salvar um arquivo CSV com os dados já limpos em seu computador.
4.  Escrever os dados numa tabela relacional em uma base de dados de sua escolha 
(MySQL, PostgreSQL, SQL Server etc.).

__Site do Ibep >> Microdados >> Enem:__


https://www.gov.br/inep/pt-br/acesso-a-informacao/dados-abertos/microdados/enem

__Url do zip do microdados do Enem-2019:__


https://download.inep.gov.br/microdados/microdados_enem_2019.zip

### Extração dos dados

#### Extrair, de maneira programática, os dados do ENEM 2019.

In [1]:
# Importar bibliotecas

import pandas as pd 
import numpy as np 
import zipfile
import requests
from io import BytesIO
import os

In [2]:
# Criar um diretório para armazenar o conteúdo do ENEM

os.makedirs('./enem2019', exist_ok=True)

In [7]:
# Define a url 
url = "https://download.inep.gov.br/microdados/microdados_enem_2019.zip"

# Download do conteúdo da url
filebytes = BytesIO(
    requests.get(url).content
)

# Descompactação do zip
myzip = zipfile.ZipFile(filebytes)

# Pasta destino da descompactação
myzip.extractall("./enem2019")

In [4]:
# Leitura dos dados

enem = pd.read_csv("./enem2019/DADOS/MICRODADOS_ENEM_2019.csv", 
                    sep = ";", 
                    decimal = "." )

In [5]:
enem

NU_INSCRICAO  NU_ANO  CO_MUNICIPIO_RESIDENCIA  \
0        190001595656    2019                  3552205   
1        190001421546    2019                  2910800   
2        190001133210    2019                  2304400   
3        190001199383    2019                  1721000   
4        190001237802    2019                  3118601   
...               ...     ...                      ...   
5095265  190006074437    2019                  5300108   
5095266  190005429225    2019                  1302603   
5095267  190006087652    2019                  4302105   
5095268  190006087653    2019                  4302105   
5095269  190006087654    2019                  4302105   

        NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  NU_IDADE  \
0                      Sorocaba                35               SP      36.0   
1              Feira de Santana                29               BA      23.0   
2                     Fortaleza                23               CE      39.0   
3                        Palmas                17               TO      25.0   
4                      Contagem                31               MG      22.0   
...                         ...               ...              ...       ...   
5095265                Bras�lia                53               DF      41.0   
5095266                  Manaus                13               AM      20.0   
5095267         Bento Gon�alves                43               RS      21.0   
5095268         Bento Gon�alves                43               RS      22.0   
5095269         Bento Gon�alves                43               RS      21.0   

        TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  ...  Q016  Q017 Q018  Q019  \
0             M                1            3  ...     A     A    A     A   
1             M                1            1  ...     A     A    A     B   
2             F                1            3  ...     A     A    A     B   
3             F                1            1  ...     A     A    A     C   
4             F                1            1  ...     B     A    A     B   
...         ...              ...          ...  ...   ...   ...  ...   ...   
5095265       F                4            3  ...     B     A    A     B   
5095266       M                1            3  ...     A     A    A     B   
5095267       M                0            0  ...     A     A    A     B   
5095268       M                0            0  ...     B     A    A     C   
5095269       F                0            0  ...     A     A    A     B   

        Q020  Q021  Q022  Q023  Q024  Q025  
0          A     A     C     A     C     B  
1          A     A     E     A     A     B  
2          A     A     C     A     A     B  
3          A     B     D     A     B     B  
4          A     A     C     B     B     B  
...      ...   ...   ...   ...   ...   ...  
5095265    A     A     B     B     A     B  
5095266    B     A     E     A     A     B  
5095267    A     A     C     B     B     B  
5095268    B     B     A     A     A     B  
5095269    A     A     B     B     B     B  

[5095270 rows x 136 columns]

#### Filtrar apenas os alunos do estado de Minas Gerais

In [6]:
dict(enem.dtypes)

{'NU_INSCRICAO': dtype('int64'),
 'NU_ANO': dtype('int64'),
 'CO_MUNICIPIO_RESIDENCIA': dtype('int64'),
 'NO_MUNICIPIO_RESIDENCIA': dtype('O'),
 'CO_UF_RESIDENCIA': dtype('int64'),
 'SG_UF_RESIDENCIA': dtype('O'),
 'NU_IDADE': dtype('float64'),
 'TP_SEXO': dtype('O'),
 'TP_ESTADO_CIVIL': dtype('int64'),
 'TP_COR_RACA': dtype('int64'),
 'TP_NACIONALIDADE': dtype('int64'),
 'CO_MUNICIPIO_NASCIMENTO': dtype('float64'),
 'NO_MUNICIPIO_NASCIMENTO': dtype('O'),
 'CO_UF_NASCIMENTO': dtype('float64'),
 'SG_UF_NASCIMENTO': dtype('O'),
 'TP_ST_CONCLUSAO': dtype('int64'),
 'TP_ANO_CONCLUIU': dtype('int64'),
 'TP_ESCOLA': dtype('int64'),
 'TP_ENSINO': dtype('float64'),
 'IN_TREINEIRO': dtype('int64'),
 'CO_ESCOLA': dtype('float64'),
 'CO_MUNICIPIO_ESC': dtype('float64'),
 'NO_MUNICIPIO_ESC': dtype('O'),
 'CO_UF_ESC': dtype('float64'),
 'SG_UF_ESC': dtype('O'),
 'TP_DEPENDENCIA_ADM_ESC': dtype('float64'),
 'TP_LOCALIZACAO_ESC': dtype('float64'),
 'TP_SIT_FUNC_ESC': dtype('float64'),
 'IN_BAIXA_VISA

In [7]:
#enemMG = enem.CO_REGIAO_CURSO == 31


enemMG = enem.loc[
    enem.CO_UF_RESIDENCIA == 31
]

In [8]:
enemMG

NU_INSCRICAO  NU_ANO  CO_MUNICIPIO_RESIDENCIA  \
4        190001237802    2019                  3118601   
14       190001237803    2019                  3170206   
25       190001237804    2019                  3124104   
32       190001237805    2019                  3118304   
38       190001237806    2019                  3162922   
...               ...     ...                      ...   
5095201  190005766803    2019                  3171204   
5095225  190005766804    2019                  3170206   
5095236  190005766805    2019                  3162500   
5095237  190005766806    2019                  3158201   
5095256  190005766807    2019                  3171303   

        NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  NU_IDADE  \
4                      Contagem                31               MG      22.0   
14                   Uberl�ndia                31               MG      30.0   
25                   Esmeraldas                31               MG      27.0   
32         Conselheiro Lafaiete                31               MG      22.0   
38         S�o Joaquim de Bicas                31               MG      23.0   
...                         ...               ...              ...       ...   
5095201              Vespasiano                31               MG      23.0   
5095225              Uberl�ndia                31               MG      28.0   
5095236        S�o Jo�o del Rei                31               MG      20.0   
5095237   Santa Maria do Sua�u�                31               MG      18.0   
5095256                  Vi�osa                31               MG      29.0   

        TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  ...  Q016  Q017 Q018  Q019  \
4             F                1            1  ...     B     A    A     B   
14            F                1            2  ...     A     A    A     B   
25            M                1            3  ...     A     A    A     B   
32            F                1            1  ...     B     A    B     C   
38            F                1            4  ...     B     A    A     B   
...         ...              ...          ...  ...   ...   ...  ...   ...   
5095201       M                1            2  ...     B     A    A     C   
5095225       F                2            1  ...     A     A    B     B   
5095236       M                1            4  ...     B     A    B     C   
5095237       F                1            3  ...     A     A    A     B   
5095256       F                1            3  ...     A     A    A     B   

        Q020  Q021  Q022  Q023  Q024  Q025  
4          A     A     C     B     B     B  
14         A     A     B     A     A     A  
25         A     B     C     A     A     A  
32         B     A     E     A     B     B  
38         B     A     C     A     B     B  
...      ...   ...   ...   ...   ...   ...  
5095201    A     B     D     A     B     B  
5095225    B     B     C     B     B     B  
5095236    B     B     C     B     B     B  
5095237    A     A     B     A     A     A  
5095256    A     A     B     A     A     A  

[538896 rows x 136 columns]

#### Salvar um arquivo CSV com os dados já limpos em seu computador.

In [9]:
enemMG.to_csv('./enem2019/DADOS/MICRODADOS_ENEM_2019_MG.csv', sep = ";", 
                    decimal = "." )

In [12]:

enemTeste = pd.read_csv("./enem2019/DADOS/MICRODADOS_ENEM_2019_MG.csv", 
                    sep = ";", 
                    decimal = "." )

In [13]:
enemTeste

Unnamed: 0  NU_INSCRICAO  NU_ANO  CO_MUNICIPIO_RESIDENCIA  \
0                4  190001237802    2019                  3118601   
1               14  190001237803    2019                  3170206   
2               25  190001237804    2019                  3124104   
3               32  190001237805    2019                  3118304   
4               38  190001237806    2019                  3162922   
...            ...           ...     ...                      ...   
538891     5095201  190005766803    2019                  3171204   
538892     5095225  190005766804    2019                  3170206   
538893     5095236  190005766805    2019                  3162500   
538894     5095237  190005766806    2019                  3158201   
538895     5095256  190005766807    2019                  3171303   

       NO_MUNICIPIO_RESIDENCIA  CO_UF_RESIDENCIA SG_UF_RESIDENCIA  NU_IDADE  \
0                     Contagem                31               MG      22.0   
1                   Uberl�ndia                31               MG      30.0   
2                   Esmeraldas                31               MG      27.0   
3         Conselheiro Lafaiete                31               MG      22.0   
4         S�o Joaquim de Bicas                31               MG      23.0   
...                        ...               ...              ...       ...   
538891              Vespasiano                31               MG      23.0   
538892              Uberl�ndia                31               MG      28.0   
538893        S�o Jo�o del Rei                31               MG      20.0   
538894   Santa Maria do Sua�u�                31               MG      18.0   
538895                  Vi�osa                31               MG      29.0   

       TP_SEXO  TP_ESTADO_CIVIL  ...  Q016  Q017  Q018 Q019  Q020 Q021  Q022  \
0            F                1  ...     B     A     A    B     A    A     C   
1            F                1  ...     A     A     A    B     A    A     B   
2            M                1  ...     A     A     A    B     A    B     C   
3            F                1  ...     B     A     B    C     B    A     E   
4            F                1  ...     B     A     A    B     B    A     C   
...        ...              ...  ...   ...   ...   ...  ...   ...  ...   ...   
538891       M                1  ...     B     A     A    C     A    B     D   
538892       F                2  ...     A     A     B    B     B    B     C   
538893       M                1  ...     B     A     B    C     B    B     C   
538894       F                1  ...     A     A     A    B     A    A     B   
538895       F                1  ...     A     A     A    B     A    A     B   

        Q023  Q024  Q025  
0          B     B     B  
1          A     A     A  
2          A     A     A  
3          A     B     B  
4          A     B     B  
...      ...   ...   ...  
538891     A     B     B  
538892     B     B     B  
538893     B     B     B  
538894     A     A     A  
538895     A     A     A  

[538896 rows x 137 columns]

#### Escrever os dados numa tabela relacional em uma base de dados de sua escolha (MySQL, PostgreSQL, SQL Server etc.).

In [26]:
!pip install pyodbc

In [34]:
!pip install sqlalchemy

In [27]:
# Ingestão no SQL Server

import pyodbc
import sqlalchemy


In [79]:
engine = sqlalchemy.create_engine(
    "mssql+pyodbc://luciana:luciana@NOBRE\SQLEXPRESS/enemMG?driver=ODBC+Driver+17+for+SQL+Server"
)

In [81]:
#conexão utilizada
engine = sqlalchemy.create_engine(
    "mssql+pyodbc://NOBRE\SQLEXPRESS/enemMG?trusted_connection=yes&driver=ODBC+Driver+13+for+SQL+Server'"
)

In [88]:
'''
from sqlalchemy import create_engine

server = 'NOBRE\SQLEXPRESS'

database = 'enemMG'

engine = create_engine('mssql+pyodbc://' + server + '/' + database + '?trusted_connection=yes&driver=ODBC+Driver+13+for+SQL+Server')
'''

In [89]:
#Importa os dados do DF para o DB
enemMG.to_sql("enemMG", con=engine, index=False, if_exists='append')

# Trabalho Prático - Questões

## 1. Qual é a média da nota em matemática de todos os alunos mineiros?

Comando SQL:


SELECT avg (NU_NOTA_MT) as [Nota média Matemática]
FROM enemMG

In [104]:
enemMG.NU_NOTA_MT.describe()

count    393872.000000
mean        546.796208
std         115.072018
min           0.000000
25%         451.800000
50%         531.400000
75%         629.800000
max         985.500000
Name: NU_NOTA_MT, dtype: float64

## Resposta: 546.8

## 2. Qual é a média da nota em Linguagens e Códigos de todos os alunos mineiros?



Comando SQL:

SELECT avg (NU_NOTA_LC) as [Nota média de Linguagens e Códigos]
FROM enemMG

In [91]:
enemMG.NU_NOTA_LC.describe()

count    414346.000000
mean        531.215550
std          61.324732
min           0.000000
25%         497.400000
50%         536.500000
75%         572.900000
max         770.500000
Name: NU_NOTA_LC, dtype: float64

## Resposta: 531.2

## 3. Qual é a média da nota em Ciências Humanas dos alunos do sexo FEMININO mineiros?

Comando SQL:

SELECT avg (NU_NOTA_CH) as [Nota média de Ciências Humanas e do sexo Feminino]
FROM enemMG
WHERE TP_SEXO LIKE 'F'

In [101]:
enemMG.loc[enemMG.TP_SEXO=='F'].NU_NOTA_CH.describe()

count    251816.000000
mean        515.127098
std          81.019828
min           0.000000
25%         458.200000
50%         519.300000
75%         572.600000
max         795.000000
Name: NU_NOTA_CH, dtype: float64

## Resposta: 515.1

## 4. Qual é a média da nota em Ciências Humanas dos alunos do sexo MASCULINO?

Comando SQL:

SELECT avg (NU_NOTA_CH) as [Nota média de Ciências Humanas e do sexo Masculino]
FROM enemMG
WHERE TP_SEXO LIKE 'M'

In [102]:
enemMG.loc[enemMG.TP_SEXO=='M'].NU_NOTA_CH.describe()

count    162530.000000
mean        529.698270
std          85.803672
min           0.000000
25%         470.200000
50%         538.500000
75%         591.500000
max         835.100000
Name: NU_NOTA_CH, dtype: float64

## Resposta: 529.7

## 5. Qual é a média da nota em Matemática dos alunos do sexo FEMININO que moram na cidade de Montes Claros?

Comando SQL:

SELECT avg (NU_NOTA_MT) as [Nota média Matemática / Feminino / Montes Claros]
FROM enemMG
WHERE TP_SEXO LIKE 'F' and NO_MUNICIPIO_RESIDENCIA LIKE 'Montes Claros'

In [103]:
enemMG.loc[(enemMG.TP_SEXO=='F') & (enemMG.NO_MUNICIPIO_RESIDENCIA=='Montes Claros')].NU_NOTA_MT.describe()

count    7699.000000
mean      525.477672
std       107.081759
min       359.300000
25%       438.750000
50%       505.400000
75%       601.000000
max       913.700000
Name: NU_NOTA_MT, dtype: float64

## Resposta: 525.5

## 6. Qual é a média da nota em Matemática dos alunos do município de Sabará que possuem TV por assinatura na residência?

Comando SQL:

SELECT avg (NU_NOTA_MT) as [Nota média Matemática / TV assinatura / Sabará]
FROM enemMG
WHERE Q021 LIKE 'B' and NO_MUNICIPIO_RESIDENCIA LIKE 'Sabar%'

In [105]:
enemMG.loc[(enemMG.CO_MUNICIPIO_RESIDENCIA==3156700) & (enemMG.Q021=='B')].NU_NOTA_MT.describe()

count    704.000000
mean     543.292756
std      115.270483
min      359.100000
25%      444.125000
50%      521.850000
75%      622.425000
max      924.700000
Name: NU_NOTA_MT, dtype: float64

## Resposta: 543.3

## 7. Qual é a média da nota em Ciências Humanas dos alunos mineiros que possuem dois fornos micro-ondas em casa?

Comando SQL: 

SELECT avg (NU_NOTA_CH) as [Nota média de Ciências Humana / +2 Fornos]
FROM enemMG
WHERE Q016 LIKE 'C'

In [107]:
enemMG.loc[enemMG.Q016=='C'].NU_NOTA_CH.describe()

count    2205.000000
mean      557.276599
std        90.127265
min         0.000000
25%       498.000000
50%       572.500000
75%       622.200000
max       775.500000
Name: NU_NOTA_CH, dtype: float64

## Resposta: 557.3

## 8. Qual é a nota média em Matemática dos alunos mineiros cuja mãe completou a pós-graduação?

Comando SQL:

SELECT avg (NU_NOTA_MT) as [Nota média Matemática / Mãe com Pós]
FROM enemMG
WHERE Q002 LIKE 'g' 

In [108]:
enemMG.loc[enemMG.Q002=='G'].NU_NOTA_MT.describe()

count    40937.000000
mean       620.007062
std        122.608027
min          0.000000
25%        525.600000
50%        629.200000
75%        710.200000
max        985.500000
Name: NU_NOTA_MT, dtype: float64

## Resposta: 620.0

## 9. Qual é a nota média em Matemática dos alunos de Belo Horizonte e de Conselheiro Lafaiete?

Comando SQL:

ELECT avg (NU_NOTA_MT) as [Nota média Matemática / BH e CL]
FROM enemMG
WHERE NO_MUNICIPIO_RESIDENCIA in ( 'Conselheiro Lafaiete', 'Belo Horizonte' ) 

In [112]:
enemMG.loc[(enemMG.NO_MUNICIPIO_RESIDENCIA=='Belo Horizonte') | (enemMG.NO_MUNICIPIO_RESIDENCIA=='Conselheiro Lafaiete')].NU_NOTA_MT.describe()

count    59238.000000
mean       578.039227
std        128.510689
min          0.000000
25%        469.700000
50%        570.500000
75%        674.200000
max        985.500000
Name: NU_NOTA_MT, dtype: float64

## Resposta: 578.0

## 10. Qual é a nota média em Ciências Humanas dos alunos mineiros que moram sozinhos?

Comando SQL:

SELECT avg (NU_NOTA_CH) as [Nota média de Ciências Humana / Moram Só]
FROM enemMG
WHERE Q005 LIKE '1'

In [113]:
enemMG.loc[enemMG.Q005==1].NU_NOTA_CH.describe()

count    10131.000000
mean       534.457339
std         86.729627
min          0.000000
25%        478.950000
50%        543.200000
75%        595.300000
max        786.600000
Name: NU_NOTA_CH, dtype: float64

## Resposta: 534.5

## 11. Qual é a nota média em Ciências Humanas dos alunos mineiros cujo pai completou pós-graduação e possuem renda familiar entre R$ 8.982,01 e R$ 9.980,00.

Comando SQL:

SELECT avg (NU_NOTA_CH) as [Nota média de Ciências Humana / Etc]
FROM enemMG
WHERE Q001 LIKE 'G' and  Q006 LIKE 'M'


In [114]:
enemMG.loc[(enemMG.Q001=='G') & (enemMG.Q006=='M')].NU_NOTA_CH.describe()

count    1118.000000
mean      586.723166
std        75.330683
min       319.300000
25%       550.425000
50%       598.450000
75%       637.525000
max       784.000000
Name: NU_NOTA_CH, dtype: float64

## Resposta: 586.7

## 12. Qual é a nota média em Matemática dos alunos do sexo Feminino que moram em Lavras e escolheram “Espanhol” como língua estrangeira?

Comando SQL:

SELECT avg (NU_NOTA_MT) as [Nota média Matemática / Feminino / Lavras / Espanhol]
FROM enemMG
WHERE TP_SEXO LIKE 'F'and NO_MUNICIPIO_RESIDENCIA LIKE 'Lavras' and TP_LINGUA LIKE '1'

In [119]:
enemMG.loc[(enemMG.TP_SEXO=='F') & (enemMG.NO_MUNICIPIO_RESIDENCIA=='Lavras') & (enemMG.TP_LINGUA==1)].NU_NOTA_MT.describe()

count    894.000000
mean     510.809508
std       94.541685
min      359.900000
25%      435.175000
50%      492.200000
75%      570.350000
max      870.200000
Name: NU_NOTA_MT, dtype: float64

## Resposta: 510.8

## 13. Qual é a nota média em Matemática dos alunos do sexo Masculino que moram em Ouro Preto?



Comando SQL:

SELECT avg (NU_NOTA_MT) as [Nota média Matemática / Masculino / Ouro Preto]
FROM enemMG
WHERE TP_SEXO LIKE 'M'and NO_MUNICIPIO_RESIDENCIA LIKE 'Ouro Preto' 

In [121]:
enemMG.loc[(enemMG.TP_SEXO=='M') & (enemMG.NO_MUNICIPIO_RESIDENCIA=='Ouro Preto')].NU_NOTA_MT.describe()

count    1230.000000
mean      555.083252
std       114.936273
min       359.000000
25%       461.250000
50%       540.500000
75%       638.650000
max       903.900000
Name: NU_NOTA_MT, dtype: float64

## Resposta: 555.1

## 14. Qual é a nota média em Ciências Humanas dos alunos surdos?


Comando SQL:

SELECT avg (NU_NOTA_CH) as [Nota média de Ciências Humana / Surdez]
FROM enemMG
WHERE IN_SURDEZ LIKE '1'

In [122]:
enemMG.loc[enemMG.IN_SURDEZ==1].NU_NOTA_CH.describe()

count    108.000000
mean     435.387963
std       61.942581
min      330.500000
25%      395.100000
50%      426.350000
75%      471.250000
max      616.500000
Name: NU_NOTA_CH, dtype: float64

## Resposta: 435.4

## 15. Qual é a nota média em Matemática dos alunos do sexo FEMININO, que moram em Belo Horizonte, Sabará, Nova Lima e Betim e possuem dislexia?

Comando SQL: 

SELECT avg (NU_NOTA_MT) as [Nota média Matemática / Etc]
FROM enemMG
WHERE NO_MUNICIPIO_RESIDENCIA in ( 'Betim', 'Belo Horizonte', 'Sabar%', 'Nova Lima' ) and TP_SEXO LIKE 'F' and IN_DISLEXIA LIKE '1' 

In [123]:
enemMG.loc[((enemMG.CO_MUNICIPIO_RESIDENCIA==3156700) | (enemMG.NO_MUNICIPIO_RESIDENCIA=='Belo Horizonte') | (enemMG.NO_MUNICIPIO_RESIDENCIA=='Nova Lima') | (enemMG.NO_MUNICIPIO_RESIDENCIA=='Betim')) & (enemMG.TP_SEXO=='F') & (enemMG.IN_DISLEXIA==1)].NU_NOTA_MT.describe()

count     31.000000
mean     582.193548
std      131.003641
min      363.600000
25%      474.250000
50%      604.400000
75%      672.750000
max      789.100000
Name: NU_NOTA_MT, dtype: float64

## Resposta: 582.2